### ISSUE TITLE
Display warnings with relative values based on low baseline values

### ISSUE THREAD

**Issue by rod-glover** [pacificclimate/plan2adapt-v2#159]
This applies to Summary and Graphs tabs, which display some values (e.g., pr, prsn) as relative change.

Feedback from Trevor:

> On the Graphs tab, if you select region Central Coast, variable Snowfall, and period Summer, the relative changes (and the range of relative changes) are enormous. This is of course because summer snowfall is negligible and any change relative to such a small baseline is relatively enormous. You can see the absolute values on the Maps tab, and of course they are zero or nearly so.
> In P2A Classic, this change shows as exactly zero for every period. I am willing to bet that is artificially set to avoid just such a misleading looking graph as the one we now see here.
> [[TQM]] I think your guess is right. This comes up a lot with regional impacts assessment reports, generally for Cooling Degree Days (which BTW would be a great one to add to the list of plots – as would SU25).
> [[TQM]]The way we handle it in those reports is a little bit of manual judgement but essentially when the baselines get small we switch from percentages to reporting absolute values. We have also had to automate this in summary tables and I think the guideline we generally use there is when % increases are > 100%
> [[TQM]] I’m not sure if we want to display total values though. In reports we have the luxury of spelling out that total values for regions need to be taken with a grain of salt and we always have specific locations to compare them to. On P2A plots we always wanted to point people to projected change so they can apply that change to any location within the region but that line of thinking breaks down when you have small baselines for a region.
> [[TQM]] Long story short I think the best way to handle it is to display a warning.  The warning we use in regional impacts analyses summary tables is: “CAUTION: Percent changes from a low baseline value can result in deceptively large percent change values”
> [[TQM]] We talked about detecting some % change threshold for this to come up but we ultimately went with just always labelling cooling degree days, SU25, and SU30 with that warning as those are always when it comes up. If you take that approach I would only label summer snow as other seasons should have reasonable baselines everywhere.

In response, I propose:
- In configuration, specify threshold for "low baseline value".
- In configuration, specify caution message.
- In data displays, add caution message for each applicable variable (as above) when baselines are low.

Implementation notes:
- Low-baseline flag can easily be added to result of function `getDisplayData`, which is used by both Summary and Graphs tabs.
- Baseline is not a single value, but a list of values, one per percentile. What is precise low baseline condition for lists?

Trevor adds: 

> Also, if you select region Capital, you will get an apparently empty graph. I believe this is because the snowfall there is zero everywhere in summer and I am dividing by a zero baseline to show relative change. This can and should be fixed, but that does not change the question about what to do about relative changes when the baseline is very nearly zero.
> [[TQM]] For region Capital I see a 70 to 100% decrease for annual but yeah with summer it’s just an empty graph. I would say you could just leave this as is. It should be pretty self explanatory. If you want to include some sort of error handling that displays a message like “plot is blank as baseline and future all contain no occurrences of snowfall in summer” that would be even better but a low priority imo

The current behaviour of the graphs with low baseline (mediated by logic in `getDisplayData`) is to show (i.e., receive) zero relative change when baseline is near zero. This logic will already be extended or replaced by the above work, which also already proposes to add a warning message.  Implementing this suggestion is essentially zero extra cost, and makes the tabs more uniform.

**Comment by rod-glover** [comment]
Hold on this. Further questions in with Trevor.

**Comment by rod-glover** [comment]
> [[TQM]] Although the error handling is easy, I don’t think it is the preferred solution. It might be hard to find something that works for the whole Province and it will definitely break once we add more variables, like SU25. There’s just no uniform criterion that works across the Province and all seasons, so I think we do need to handle it with a blanket caveat. I’m a fan of this: “CAUTION: Percent changes from a low baseline value can result in deceptively large percent change values”. Right now we would only display that for Summer Snowfall. If Cooling Degree Days were displayed as percent changes we’d need it there too but I think showing those changes in degree-days works.
> [[REG]] Just to clarify our capabilities here: We can implement a wide range of "low baseline" computations, differentiated by variable and other criteria (e.g., season, region) if necessary. So when you say "uniform criterion" it need not be uniform by variable, season or even region. Would you consider such a thing, or is this still infeasible?
> [[REG]] Supposing that even distinct criteria per variable, etc., will not serve, can I confirm that you want a caution message as above displayed for each relevant variable (so far, prsn)? And if so, how is the user to know that a low baseline condition prevails?
> 
> [[TQM]] The problem isn’t what you can implement, it’s that we can’t give you a coherent set of criteria that will cover all cases. And the harder we try / the closer we get to covering all cases, the more confusing it will be when people encounter cases we inevitably miss. I’d rather leave the caution which is unsatisfactory in the sense that people might be left with the question of “but what does a low baseline mean in this context and what even is the baseline”. I’d rather get that question once in a while and educate the user than end up with people getting misled by an impossible-to-make-perfect set of criteria.
> 
> [[REG]] OK, understood now that specifying precise computational criteria is not practical -- too many special cases. 
> [[REG]] That said, how is the user to know that a low baseline condition prevails?​ What clues will they have that this caution is actually in force for any given selection of region, future time period, variable, season? In none of the tabs (specifically, Summary, Graphs) is there an indication of what the baseline values are. A user would have to go to the maps, possibly select the right future time period, and estimate values in their region of interest. That seems a bit of a high expectation for users of P2A, or really even anyone.

It looks as if the caution will be present without computational conditions or context information. To be continued...

**Comment by rod-glover** [comment]
> [[REG]] OK, understood now that specifying precise computational criteria is not practical -- too many special cases. 
> [[TQM]] I think so, but that’s only my (somewhat well informed) intuition. I’m very much open to being proven wrong but I suspect this could be a bit of a time intensive rabbit hole.
>  
> [[REG]] That said, how is the user to know that a low baseline condition prevails?​
> [[TQM]] They won’t, for now. Not until PRISM is included and we can start sprinkling baseline information in places. But even then, it will still be a judgement call. Most users won’t know a low baseline even when it’s displayed. Most experts won’t agree on the dividing line between what is and isn’t a low baseline. I might not even agree with myself on different days of the week.
>  
> What clues will they have that this caution is actually in force for any given selection of region, future time period, variable, season?
> [[TQM]] The only clues are (1) the presence of this warning, and (2) maybe the presence of large percentage changes.
>  
> In none of the tabs (specifically, Summary, Graphs) is there an indication of what the baseline values are. A user would have to go to the maps, possibly select the right future time period, and estimate values in their region of interest. That seems a bit of a high expectation for users of P2A, or really even anyone.
> [[TQM]] I’m pretty comfortable with users having to dig a bit and give this issue some thought. The concept that not all large relative changes are created equally and some are just math artefacts of low baselines is a concept that everyone has to learn on their journey. I’m ok them scratching their heads a bit on this because it’s for a good reason that will lead to learning. I guess I’m saying confusing is better than misleading especially when the confusion leads to a good lesson. This is very much a subjective point that is open to criticism.

From Kari:

> I’m going to chime in with affirmation the point that: confusing is better than misleading especially when the confusion leads to a good lesson.
>  
> I’ve been doing a lot of compare and contrast on climate tools lately and I think that is a consistent  design intent behind P2A. There are a couple of examples of climatedata.ca addressing confusion in ways that, I think, are misleading and are leading to much more challenging conversations.
>  
> It’s much easier to bring people along on a learning journey when we’re giving people advice on how to find additional information, or think creatively about what is available, than it is to tell them that they’ve been doing something wrong, when it felt intuitive or right to them.

So we have settled on a simple approach, which is that some variables (currently, prsn) will come with a warning reading:

CAUTION: Percent changes from a low baseline value can result in deceptively large percent change values. This variable may have low baseline values.

The wording of the warning, and its relevance to any given variable will be derived from the config file.
